### ISSUE TITLE
Chef 12 node attributes missing when using Ruby ":" string syntax instead of quotes

### ISSUE THREAD

**Issue by eric-tucker** [chef/chef#2712]
This is possibly related to Chef Zero and/or force_default.  I am using the community solr cookbook but I need to override attributes it sets.

I have these solr attributes overridden in my attribute file.
````
force_default[:solr][:version] = '4.10.2'
force_default[:solr][:url] = "https://archive.apache.org/dist/lucene/solr/#{node[:solr][:version]}/solr-#{node[:solr][:version]}.tgz"
force_default[:solr][:port] = 8983
force_default[:solr][:data_dir] = "/opt/solr-#{node[:solr][:version]}/example/solr"
force_default[:solr][:xms] = "512M"
force_default[:solr][:xmx] = "1536M"
````

If I try to access node attributes using `node[:solr][:some_attribute]` they are all missing except version.

I decided to pretty print the hash values using the quote string notiation vs the colon notation:
`pp node[:solr]`
````
{"version"=>"4.10.2"}
````
`pp node["solr"]`
````
{"version"=>"4.10.2",
 "url"=>"https://archive.apache.org/dist/lucene/solr/4.10.2/solr-4.10.2.tgz",
 "data_dir"=>"/opt/solr-4.10.2/example/solr",
 "dir"=>"/opt/solr",
 "port"=>8983,
 "pid_file"=>"/var/run/solr.pid",
 "log_file"=>"/var/log/solr.log",
 "install_java"=>true,
 "xms"=>"512M",
 "xmx"=>"1536M"}
````
Note that using Chef 11.x this behaviour does not present itself.

Was something changed around attribute handling for quotes vs colon string notation?

**Comment by AlexClineBB** [comment]
I'm also running into a variation of this issue. For me, attribute overrides that used to work in chef 11 using either quotes or colon notation no longer work in chef 12.

For example, this worked in chef 11

    # Community cookbook foobar sets the default attribute
    default.foobar[:dir] = "/usr/local"

    # I override the attribute with a new path in a wrapper cookbook
    default['foobar']['dir'] = "/opt"

   # Output the values for debugging
    puts node.foobar[:dir] # /opt
    puts default['foobar']['dir'] # /opt

    # The directory "/opt" is created by the community cookbook
    directory node.foobar[:dir] do
      action :create
    end

In chef 12, I get the following

    # Community cookbook foobar sets the default attribute
    default.foobar[:dir] = "/usr/local"

    # I override the attribute with a new path in a wrapper cookbook
    default['foobar']['dir'] = "/opt"

    # Output the values for debugging
    puts default.foobar[:dir]  # "/usr/local"
    puts default['foobar']['dir']  # "/opt"

    # The directory "/usr/local" is created by the community cookbook
    directory node.foobar[:dir] do
      action :create
    end

It seems like chef 12 is referencing two objects when different notation is used.

**Comment by eric-tucker** [comment]
Just confirmed this is NOT related to Chef Zero.  This happens with Chef Server as well.

**Comment by AlexClineBB** [comment]
I'm using chef-client 11.16.4 and 12.0.3 with chef-server.

**Comment by tyler-ball** [comment]
I'm having trouble repro-ing this.  I have the following cookbook that I'm running with `chef-client --local-mode`:

_attributes/default.rb_
```ruby
default.foobar[:dir] = "/usr/local"
default['foobar']['dir'] = "/opt"

puts default.foobar[:dir]
puts default['foobar']['dir']
```

_recipes/default.rb_
```ruby
puts node.default.foobar[:dir]
puts node.default['foobar']['dir']
```

When I run this recipe in local_mode, I see the following output on STDOUT:

```bash
Compiling Cookbooks...
/opt
/opt
/opt
/opt
Converging 0 resources
```

What should my simple repro cookbook setup look like?  This was with 12.0.3

**Comment by eric-tucker** [comment]
I'm not entirely sure.  The odd thing is I haven't noticed this elsewhere - it's only with my solr hash of attributes.  I can't seem to reproduce with any others.  But I can't see this being specific to a certain cookbook.   We're just talking about attribute precedence and compilation here.

Keep in mind I'm trying to force default so maybe it's a precedence issue.

I've tried this right in my attribute file where I do the force_default definitions:

````
force_default[:solr][:version] = '4.10.2'
force_default[:solr][:url] = "https://archive.apache.org/dist/lucene/solr/#{node[:solr][:version]}/solr-#{node[:solr][:version]}.tgz"
force_default[:solr][:port] = 8983
force_default[:solr][:data_dir] = "/opt/solr-#{node[:solr][:version]}/example/solr"
force_default[:solr][:xms] = "512M"
force_default[:solr][:xmx] = "1536M"

pp node[:solr]
pp node["solr"]
````
Here is the STDOUT I catch:
````
{"version"=>"4.10.2"}
{"version"=>"4.10.2",
 "url"=>"https://archive.apache.org/dist/lucene/solr/4.10.2/solr-4.10.2.tgz",
 "port"=>8983,
 "data_dir"=>"/opt/solr-4.10.2/example/solr",
 "xms"=>"512M",
 "xmx"=>"1536M"}
````

Now THIS is bizarre though:
````
force_default["solr"][:version] = '4.10.2'
force_default["solr"][:url] = "https://archive.apache.org/dist/lucene/solr/#{node["solr"][:version]}/solr-#{node["solr"][:version]}.tgz"
force_default["solr"][:port] = 8983
force_default["solr"][:data_dir] = "/opt/solr-#{node["solr"][:version]}/example/solr"
force_default["solr"][:xms] = "512M"
force_default["solr"][:xmx] = "1536M"

pp node[:solr]
pp node["solr"]
````
Notice that I changed all the force_default lines to use quotes instead of colons.  Here is the output:
````
{"version"=>"4.10.2",
 "url"=>"https://archive.apache.org/dist/lucene/solr/4.10.2/solr-4.10.2.tgz",
 "port"=>8983,
 "data_dir"=>"/opt/solr-4.10.2/example/solr",
 "xms"=>"512M",
 "xmx"=>"1536M"}
{"version"=>"4.10.2"}
````
It's now REVERSED.  :solr has all the attributes while "solr" has only version.

None of this makes any sense.  This is probably related to #2700.

**Comment by tyler-ball** [comment]
@eric-tucker It is something specific to `force_default` and `force_override` - When I try your example with `default` or `override` I see the correct output:

```
{"version"=>"4.10.2",
 "url"=>"https://archive.apache.org/dist/lucene/solr/4.10.2/solr-4.10.2.tgz",
 "port"=>8983,
 "data_dir"=>"/opt/solr-4.10.2/example/solr",
 "xms"=>"512M",
 "xmx"=>"1536M"}
{"version"=>"4.10.2",
 "url"=>"https://archive.apache.org/dist/lucene/solr/4.10.2/solr-4.10.2.tgz",
 "port"=>8983,
 "data_dir"=>"/opt/solr-4.10.2/example/solr",
 "xms"=>"512M",
 "xmx"=>"1536M"}
```

So I'm marking this as definitely a bug.  Can you use `override` attributes to get around this problem until we find a solution?

**Comment by tsilen** [comment]
I noticed that this would seems to happen also with "default" when strings are used rather than symbols.

In attributes file  I had something like:
```
default["foo"]["bar"] = "xyz"
```

and those attributes were no longer available as node.foo.bar or node[:foo][:bar] in recipe when the client was upgraded from 11 to 12.

Changing them to 

```
default[:foo][:bar] = "xyz"
```

seemed to fix the problem.

In my case the client is 12.0.3 and server is still 11.x.

**Comment by lamont-granquist** [comment]
closed by #2753
